src/replay/confirm_slot.zig

```diff
@@ -0,0 +1,548 @@
+        return .{
+            switch (self.poh_verifier.pollFallible()) {
+                .done => .done,
+                .pending => .pending,
+                .err => .{ .err = .{ .invalid_block = .InvalidEntryHash } },
+            },
+            try self.scheduler.poll(),
+        };
+    }
+};
+
+const PohTask = struct {
+    allocator: Allocator,
+    initial_hash: Hash,
+    entries: []const Entry,
+
+    pub fn run(self: *PohTask) !void {
+        if (!try core.entry.verifyPoh(self.entries, self.allocator, null, self.initial_hash)) {
+            return error.PohVerifyFailed;
+        }
+    }
+};
+
+pub const VerifyTicksParams = struct {
+    /// epoch-scoped constant
+    hashes_per_tick: ?u64,
+
+    // slot-scoped constants
+    slot: u64,
+    max_tick_height: u64,
+
+    // slot-scoped state (constant during lifetime of this struct)
+    tick_height: u64,
+    slot_is_full: bool,
+
+    /// slot-scoped state (expected to be mutated while verifying ticks)
+    tick_hash_count: *u64,
+};
+
+/// Verify that a segment of entries has the correct number of ticks and hashes
+/// analogous to [verify_ticks](https://github.com/anza-xyz/agave/blob/161fc1965bdb4190aa2d7e36c7c745b4661b10ed/ledger/src/blockstore_processor.rs#L1097)
+fn verifyTicks(
+    logger: ScopedLogger,
+    entries: []const Entry,
+    params: VerifyTicksParams,
+) ?BlockError {
+    const next_bank_tick_height = params.tick_height + core.entry.tickCount(entries);
+    const max_bank_tick_height = params.max_tick_height;
+
+    if (next_bank_tick_height > max_bank_tick_height) {
+        logger.warn().logf("Too many entry ticks found in slot: {}", .{params.slot});
+        return .TooManyTicks;
+    }
+
+    if (next_bank_tick_height < max_bank_tick_height and params.slot_is_full) {
+        logger.info().logf("Too few entry ticks found in slot: {}", .{params.slot});
+        return .TooFewTicks;
+    }
+
+    if (next_bank_tick_height == max_bank_tick_height) {
+        if (entries.len == 0 or !entries[entries.len - 1].isTick()) {
+            logger.warn().logf("Slot: {} did not end with a tick entry", .{params.slot});
+            return .TrailingEntry;
+        }
+
+        if (!params.slot_is_full) {
+            logger.warn().logf("Slot: {} was not marked full", .{params.slot});
+            return .InvalidLastTick;
+        }
+    }
+
+    const hashes_per_tick = params.hashes_per_tick orelse 0;
+    if (!core.entry.verifyTickHashCount(entries, logger, params.tick_hash_count, hashes_per_tick)) {
+        logger.warn().logf("Tick with invalid number of hashes found in slot: {}", .{params.slot});
+        return .InvalidTickHashCount;
+    }
+
+    return null;
+}
+
+/// Analogous to [BlockstoreProcessorError](https://github.com/anza-xyz/agave/blob/161fc1965bdb4190aa2d7e36c7c745b4661b10ed/ledger/src/blockstore_processor.rs#L779)
+pub const ConfirmSlotError = union(enum) {
+    failed_to_load_entries: anyerror,
```

Comment:
Can we avoid anyerror here? If this is some generic error that encompasses too much to describe concretely, I don't think it's worth propagating.

Reply:
This stuff is going to be used in my next task, which is processing the replay results. I'll switch to a string now, which will probably be fine, and I'll sort it out later.

28311c8cd58b191fd67a2ce8dff8e6efa4cbd517